File: erlang/meetup/meetup.erl
-module(meetup).
-export([schedule/4]).

schedule(Year, Month, Weekday, teenth) ->
    teenthday(Year, Month, weekday_to_integer(Weekday));
schedule(Year, Month, Weekday, first) ->
    nth(Year, Month, weekday_to_integer(Weekday), 1);
schedule(Year, Month, Weekday, second) ->
    nth(Year, Month, weekday_to_integer(Weekday), 2);
schedule(Year, Month, Weekday, third) ->
    nth(Year, Month, weekday_to_integer(Weekday), 3);
schedule(Year, Month, Weekday, fourth) ->
    nth(Year, Month, weekday_to_integer(Weekday), 4);
schedule(Year, Month, Weekday, last) ->
    last(Year, Month, weekday_to_integer(Weekday)).

%% Find the first occurrance of the weekday in the month and then add a number
%% of weeks. This is fine because all months have at least 28 days!
nth(Year, Month, Weekday, N) ->
    D = first_day(Year, Month, Weekday, lists:seq(1, 7)),
    Offset = 7 * (N - 1),
    {Year, Month, D + Offset}.

teenthday(Year, Month, Weekday) ->
    Date = first_day(Year, Month, Weekday, lists:seq(13, 19)),
    {Year, Month, Date}.

%% Walk backwards from the end of the month until we hit the desired weekday.
last(Year, Month, Weekday) ->
    Date = first_day(Year, Month, Weekday, last_days(Year, Month)),
    {Year, Month, Date}.

%% Find the first day of a given week
first_day(Year, Month, Weekday, Dates) -> 
    first(fun (D) ->
                  Weekday =:= calendar:day_of_the_week({Year, Month, D})
          end, Dates).

%% Detect first item that matches Fn. This probably exists maybe?
first(_, []) ->
    {error, nomatch};
first(Fn, [H|T]) ->
    case apply(Fn, [H]) of
        true -> H;
        false -> first(Fn, T)
    end.

%% Last 7 days of each month
last_days(Y, M) ->
    LastDay = calendar:last_day_of_the_month(Y, M),
    lists:seq(LastDay, (LastDay - 7), -1).
    
%% Data: weekday symbol to numeric
weekday_to_integer(monday) -> 1;
weekday_to_integer(tuesday) -> 2;
weekday_to_integer(wednesday) -> 3;
weekday_to_integer(thursday) -> 4;
weekday_to_integer(friday) -> 5;
weekday_to_integer(saturday) -> 6;
weekday_to_integer(sunday) -> 7.

File: scala/meetup/src/main/scala/Meetup.scala
import java.time.temporal.ChronoUnit
import java.time.{DayOfWeek, LocalDate, YearMonth}

case class Meetup(month: Int, year: Int) {
  import Meetup._

  val daysInCurrentMonth: Int =
    YearMonth.of(year, month).lengthOfMonth()

  def teenth(day: DayOfWeek): LocalDate = {
    val begin = LocalDate.of(year, month, 13)
    iterateFrom(begin).find(d => d.getDayOfWeek == day).get
  }

  def first(day: DayOfWeek): LocalDate = {
    val begin = LocalDate.of(year, month, 1)
    iterateFrom(begin).find(d => d.getDayOfWeek == day).get
  }

  def second(day: DayOfWeek): LocalDate =
    first(day).plus(7, ChronoUnit.DAYS)

  def third(day: DayOfWeek): LocalDate =
    first(day).plus(14, ChronoUnit.DAYS)

  def fourth(day: DayOfWeek): LocalDate =
    first(day).plus(21, ChronoUnit.DAYS)

  def last(day: DayOfWeek): LocalDate = {
    val begin = LocalDate.of(year, month, daysInCurrentMonth)
    iterateFrom(begin, -1).find(d => d.getDayOfWeek == day).get
  }
}

object Meetup {
  val Mon = DayOfWeek.MONDAY
  val Tue = DayOfWeek.TUESDAY
  val Wed = DayOfWeek.WEDNESDAY
  val Thu = DayOfWeek.THURSDAY
  val Fri = DayOfWeek.FRIDAY
  val Sat = DayOfWeek.SATURDAY
  val Sun = DayOfWeek.SUNDAY

  private def iterateFrom(start: LocalDate, amount: Int = 1): Iterator[LocalDate] = {
    Iterator.iterate(start)(d => d.plus(amount, ChronoUnit.DAYS))
  }
}
